### PR TITLE
create webElement wrapper and Driver

### PR DESCRIPTION
**Pull request by TarasTsyhanko**
I've created webElement Wrapper and DriverFacrory

### CODE REVIEW

File: src/main/java/config/ConfigConstants.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package config;
+
+public class ConfigConstants {
```
Comment (MykhailoTsiganko) [inline src/main/java/config/ConfigConstants.java]: remove

File: src/main/java/factory/DriverProvider.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package factory;
+
+import org.openqa.selenium.SearchContext;
+import org.openqa.selenium.WebDriver;
+
+@FunctionalInterface
+public interface DriverProvider {
+    WebDriver get(WebDriver diver);
```
Comment (MykhailoTsiganko) [inline src/main/java/factory/DriverProvider.java]: use SearchContext interface instead WebDriver

File: src/main/java/page/AbstractBasePage.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import decorator.ElementDecorator;
+import factory.DriverFactory;
+import factory.DriverProvider;
+import org.openqa.selenium.support.PageFactory;
+
+public class AbstractBasePage {
+    private DriverProvider provider;
+    protected AbstractBasePage(){
+        provider = driver->driver ;
+        PageFactory.initElements(new ElementDecorator(provider.get(DriverFactory.getDiver())),this);
```
Comment (MykhailoTsiganko) [inline src/main/java/page/AbstractBasePage.java]: here you should pass not driver itserlf but driverProvider
becouse it does't make any difference from case wen you passing driver

Comment (MykhailoTsiganko) [reply src/main/java/page/AbstractBasePage.java]: PageFactory.initElements(new ElementDecorator(provider),this)
for this you should create custom element decorator class

**Review by MykhailoTsiganko — CHANGES_REQUESTED** [review]

File: src/main/java/page/AbstractBasePage.java (hunk 3 above)
Comment (TarasTsyhanko) [reply src/main/java/page/AbstractBasePage.java]: new ElementDecorator - this is my class wich decorete PageElement

Comment (TarasTsyhanko) [reply src/main/java/page/AbstractBasePage.java]: first comment i don't undestand

Comment (MykhailoTsiganko) [reply src/main/java/page/AbstractBasePage.java]: > new ElementDecorator - this is my class wich decorete PageElement

You should create Your own ElementDecorator that will take driverProvider as a parametr

File: src/main/java/factory/DriverProvider.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package factory;
+
+import org.openqa.selenium.SearchContext;
+import org.openqa.selenium.WebDriver;
+
+@FunctionalInterface
+public interface DriverProvider {
+    SearchContext get(WebDriver diver);
```
Comment (MykhailoTsiganko) [inline src/main/java/factory/DriverProvider.java]: get shouln't take any params

File: src/main/java/decorator/ElementDecorator.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import org.openqa.selenium.support.FindBys;
+import org.openqa.selenium.support.pagefactory.DefaultElementLocatorFactory;
+import org.openqa.selenium.support.pagefactory.DefaultFieldDecorator;
+import org.openqa.selenium.support.pagefactory.ElementLocator;
+
+
+public class ElementDecorator extends DefaultFieldDecorator {
+
+    public ElementDecorator(DriverProvider provider) {
+        super(new DefaultElementLocatorFactory(provider.get(DriverFactory.getDiver())));
```
Comment (MykhailoTsiganko) [inline src/main/java/decorator/ElementDecorator.java]: create you custom ElementLocatorFactory that will take provider as a parametr

File: src/main/java/decorator/LocatingCustomElementListHandler.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+import elements.PageElement;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.pagefactory.ElementLocator;
+
+
+public class LocatingCustomElementListHandler implements InvocationHandler {
+	private final ElementLocator locator;
```
Comment (MykhailoTsiganko) [inline src/main/java/decorator/LocatingCustomElementListHandler.java]: you should create Your own element locator factory

**Review by MykhailoTsiganko — CHANGES_REQUESTED** [review]

**Pull request merged**